### PR TITLE
Use `thiserror` derive crate for `std::error::Error` implementation

### PR DESCRIPTION
**Pull request by yoep**
This PR implements some small convenience features for users and maintainers.

- Replaced the `std::error::Error` implementation with the `Error` derive macro from `thiserror` crate
- Added new feature `cast` for exporting the `cast` package for testing purposes
- Removed `panic` workaround statements within tests because of the missing `PartialEq` implementations

### CODE REVIEW

File: src/errors.rs [hunk 1]
```diff
@@ -1,62 +1,41 @@
...
-    io::Error as IoError,
-};
+use thiserror::Error;
 
 /// Consolidates possible error types that can occur in the lib.
-#[derive(Debug)]
+#[derive(Debug, Error)]
 pub enum Error {
     /// This variant is used when error occurs in the lib logic.
+    #[error("an internal error occurred, {}", _0)]
```
Comment (azasypkin) [inline src/errors.rs]: __question:__ wouldn't this be a bit more readable (here and in other cases below)?
```suggestion
    #[error("an internal error occurred, {0}")]
```

**Review by azasypkin — APPROVED** [review]
Nice change, thanks! Just left one optional question\nit you might want to address before we merge.

File: src/errors.rs (hunk 1 above)
Comment (yoep) [reply src/errors.rs]: Oh wow, I didn't notice that was also a valid option within `thiserror` crate.
I'll update it with the proposed change.

**Comment by yoep** [comment]
The new cargo version seems to suddenly be a lot more strict on `dead_code` within structs than in the past due to sudden new clippy failures.
As a workaround, I've annotated the structs with `#[allow(dead_code)]`

**Comment by azasypkin** [comment]
> The new cargo version seems to suddenly be a lot more strict on dead_code within structs than in the past due to sudden new clippy failures.
As a workaround, I've annotated the structs with #[allow(dead_code)]

Ha, interesting.... Everything looks good, thanks again! Merging.

**Pull request merged**